rw.go

```diff
@@ -4,95 +4,111 @@ import (
 	"encoding/binary"
 	"fmt"
 	"io"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"golang.org/x/crypto/poly1305"
 )
 
-// Each encrypted transport message must be <= 65,535 bytes, including 16
-// bytes of authentication data. To write larger plaintexts, we split them
-// into fragments of maxPlaintextLength before encrypting.
-const maxPlaintextLength = 65519
+// MaxTransportMsgLength is the Noise-imposed maximum transport message length,
+// inclusive of the MAC size (16 bytes, Poly1305 for noise-libp2p).
+const MaxTransportMsgLength = 0xffff
+
+// MaxPlaintextLength is the maximum payload size. It is MaxTransportMsgLength
+// minus the MAC size. Payloads over this size will be automatically chunked.
+const MaxPlaintextLength = MaxTransportMsgLength - poly1305.TagSize
 
-// Read reads from the secure connection, filling `buf` with plaintext data.
-// May read less than len(buf) if data is available immediately.
+// LengthPrefixLength is the length of the length prefix itself, which precedes
+// all transport messages in order to delimit them. In bytes.
+const LengthPrefixLength = 2
+
+// Read reads from the secure connection, returning plaintext data in `buf`.
+//
+// Honours io.Reader in terms of behaviour.
 func (s *secureSession) Read(buf []byte) (int, error) {
 	s.readLock.Lock()
 	defer s.readLock.Unlock()
 
-	l := len(buf)
-
-	// if we have previously unread bytes, and they fit into the buf, copy them over and return
-	if l <= len(s.msgBuffer) {
-		copy(buf, s.msgBuffer)
-		s.msgBuffer = s.msgBuffer[l:]
-		return l, nil
+	// 1. If we have queued received bytes:
+	//   1a. If len(buf) < len(queued), saturate buf, update seek pointer, return.
+	//   1b. If len(buf) >= len(queued), copy remaining to buf, release queued buffer back into pool, return.
+	//
+	// 2. Else, read the next message off the wire; next_len is length prefix.
+	//   2a. If len(buf) >= next_len, copy the message to input buffer (zero-alloc path), and return.
+	//   2b. If len(buf) < next_len, obtain buffer from pool, copy entire message into it, saturate buf, update seek pointer.
+	if s.qbuf != nil {
+		// we have queued bytes; copy as much as we can.
+		copied := copy(buf, s.qbuf[s.qseek:])
+		if s.qseek += copied; s.qseek == len(s.qbuf) {
```

Comment:
nit:
```suggestion
		s.qseek += copied
		if s.qseek == len(s.qbuf) {
```

(it has a side effect)

Reply:
Mood killer.